docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,281 @@
+}
+```
+
+### Mirror node polling:
+
+1. Maintain a list of all live socket connections. This can be done with a `key-value` store where the `key` is a unique identifier (some ws libraries provide such an id, but if needed it can be generated by the relay) of the socket object, and the `value` is the socket object itself.
+
+2. Maintain a list of all active subscriptions and the corresponding socket objects. It should have the following structure:
+
+```javascript
+  subscriptions: [
+    {
+      method: string, // newHeads, logs, etc
+      params: any,
+      subscribers: [{   // a list of all active subscriptions to the specific `method + params` combination
+        socketId: string,
+        subscription: string
+      }],
+      lastUpdated: BlockNumber // The latest block number at which this subscription was updated
+    }
+  ]
+
+```
+
+3. Create a scheduled job that loops through all saved subscriptions and calls the Mirror node Rest API and sends the response to all `subscribers`. It should be possible to configure the time interval. A throttling mechanism can be implemetented that limits the consecutive calls to the Mirror node, and at every interval only X subscriptions with the oldest `lastUpdated` should be polled.
+
+## Error Codes
+
+| Error Codes |                                                       Error message                                                        |                                    Solution                                     |
+| :---------: | :------------------------------------------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------: |
+|    32600    | "The maximum batch size that can be sent over a websocket connection is 10. Please decrease the batch size and try again.” | Occurs when user attempts to send high-volume JSON-RPC traffic over Websockets. |
+
+## Limits:
+
+1. WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.
+
+## Tasks (in suggested order):
+
+#### Milestone 1 (done)
+
+1. Finalize design document
+
+#### Milestone 2
+
+1. Implement library for websocket support in the server.
+2. Extend current koa server logic.
+3. Introduce `eth_subscribe` and `eth_unsubscribe` to both server and relay part.
+4. Add logic to save all subsciption IDs and it's corresponding filter to a cache.
+5. Add scheduled job function for looping through all saved filters and returns result to it's corresponding socket with the respective subscription id.
+6. Extend current ratelimiter class to support websocket limitation.
+7. Add logic to delete subsciption upon connection break or request.
+
+#### Milestone 3 (use Mirror-Node GraphQL API)
```

Comment:
Note that the Mirror Node has not yet design this feature at the time of writing and there is no current schedule for when the relay would adopt this